Add ability for users to include arbitrary files within the EFI partition

DjLogozzo opened the pull request:
#### Edit: Focus of this PR has changed since it was created, the old description can be seen below

This PR adds the ability for users to include arbitrary files within the FAT image file created. This functionality is controlled with the `--add_files <location_on_disk>:<location_within_image>` argument.
This flag can be supplied  multiple times to include multiple files, and the `<location_within_image>` can be excluded (eg: `--add_file <location_on_disk>`) to have the file be placed at the root of the image with the name of the exterior file.


#### Old Description

Converted the startup method for the .efi app from a 'startup.nsh' script into a boot entry saved in non-volatile ram (NvVars file)
This method has the benefit of not having a 5 second wait at the start of every boot cycle

The steps required to create the NvVars serialised binary file which contains the run.efi file as boot 0 are
- Start uefi-run with a standard .efi file to create an image.fat file
	- `$ uefi-run /Path/To/.efi`
- Copy the image.fat file from the temporary directory and save it somewhere else, (your image.fat may not be stored in this dir, it depends on where the tempfile crate puts temp files on your system)
	- `$ cp /tmp/.tmp######/image.fat /Path/To/image.fat`
- Close uefi-run and start a qemu session with the following command, these steps are done to create a persistent session that we can extract the NvVars file from
	- `$ qemu-system-x86_64 -bios /usr/share/ovmf/OVMF.fd -net none -drive file=/Path/To/image.fat,index=0,media=disk,format=raw`
- Once qemu launched, press escape to stop the auto-startup
- Inside the uefi shell, type the following command, this command will create the boot option, set it to priority 0, and save it to the NvVars file (It is saved to a file as this is the backup system uefi uses when a system does not support non-volatile ram)
	- `Shell> bcfg boot add 0 FS0:\run.efi "uefi_boot"`
- Close qemu
- Mount the image.fat file to a directory (this step requires root access)
	- `$ mkdir /Path/To/Temp/Dir`
	- `# mount /Path/To/image.fat /Path/To/Temp/Dir`
- Navigate to the mounted image.fat filesystem
	- `$ cd /Path/To/Temp/Dir`
- Copy the 'NvVars' file out to a safe location
	- `$ cp ./NvVars /Path/To/Somewhere/Safe`
- You can now unmount the image.fat and delete the temp dir (this step requires root access)
	- `# umount /Path/To/Temp/Dir`
	- `$ rmdir /Path/To/Temp/Dir`

DjLogozzo committed:
Add the ability to include arbitrary files in the FAT Image

This PR adds the ability for users to include arbitrary files within the FAT image file created.
This functionality is controlled with the `--add_files <location_on_disk>:<location_within_image>` argument.
This flag can be supplied multiple times to include multiple files,
and the `<location_within_image>` can be excluded (eg: `--add_file <location_on_disk>`) to have the file
be placed at the root of the image with the name of the exterior file.

Richard-W commented:
Hi, thanks for your contribution.

I can't accept this without changes though; sorry. `uefi-run` is supposed to "just work" after installation. No further setup should be needed to run EFI executables.

If I understand your comments correctly you propose that to get uefi-run to run correctly you need to:
* Compile uefi-run
* Follow some manual instruction to create NvVars
* Mount partition image and extract NvVars file
* Recompile uefi-run with new NvVars file

This seems to be a very complicated approach to solve a simple problem (which is a 5 second wait time). I am willing to accept a PR which adds support for adding files (including, but not limited to NvVars) to the EFI partition via a command-line option. That should solve your problem too, right?

Best regards,
Richard

PS: Please run `cargo fmt` on your commits as the build robot will decline any changes that differ from that formatting.

DjLogozzo commented:
The NvVars file should already be included in my commit as does just-work for me when I clone my fork with the commit to a fresh directory and run `cargo run run.efi`
The steps I included was purely for documentation purposes so you would understand where I got the serialised binary NvVars file as I personally wouldn't trust an arbitrary binary file appearing in a pull request to my repo with no apparent source

The steps required to get uefi-run to work with my commit should be the exact same as they are currently, as my intention was to streamline the usage of the application and the boot process as much as possible.

The steps required to get this commit working for me are as follows
- `git clone https://github.com/DjLogozzo/uefi-run`
- `cd uefi-run`
- `cargo run /Path/To/run.efi`
Or alternatively
- `cargo install`
- `uefi-run /Path/To/run.efi`

Apologies for not making that clearer in my initial pull request, i can see how my comment suggests a much longer setup process

Richard-W commented:
I still see some problems with this. I don't think UEFI specifies the exact binary format of the NvVars file so it may differ between
* OVMF versions
* Other bios versions supported via the `--bios` switch

I still think adding a switch like `--add-file /path/on/host/system:/path/on/efi/partition` would be better since it doesn't require binary files in this repository. Even if I trust you to do generate it correctly - other people may not.

DjLogozzo commented:
That is a fair concern, I didn't even think about whether the NvVars file was consistent across UEFI implimentations.

Ive refactored the code to allow for arbitrary file inclusion within the FAT filesystem. Let me know if you would like the wording of the argument option or help text altered.

The way I've implemented it, the file on system and the file on efi partition are passed as two separate strings (space separated) so you would pass it as `--add-file /Path/On/System /Path/Within/Image` (the argument settings I've set requires 2 args for the option, so clap will complain and halt the program if only one is passed or three or more are passed per invocation).
If you would prefer it to be colon separated `--add-file /Path/On/System:/Path/Within/Image` as it is suggested in your comment, I can change it to that

DjLogozzo commented:
I've converted the way that users pass files from `<outer file> <inner file>` to `<outer file>:<inner file>` like it is in your suggestion, this allows us to default the name of the inner file to the name of the outer file placed in the root if the user does not provide an inner name

richard-w commented on src/main.rs:
```diff
@@ -125,6 +140,49 @@ fn main() {
             .unwrap();
+
+        // Create user provided additional files
+        for files in additional_files {
```
```suggestion
        for file in additional_files {
```
Use singular form in iteration

richard-w commented on src/main.rs:
```diff
@@ -125,6 +140,49 @@ fn main() {
+
+            let mut files = files.split(':');
+            // Safe to unwrap, there will never be an empty string passed as an arg
+            let outer = files.next().unwrap();
```
```suggestion
            let outer = PathBuf::from(file.next().expect("Invalid --add-file argument"));                                                                                                                         
            let inner = PathBuf::from(file.next().expect("Invalid --add-file argument"));
```
Verify the whole argument early. Also use `PathBuf` to represent paths.

DjLogozzo replied:
Good call to use PathBuf, it seems I've spent far too much time recently within #![no_std] because I completely forgot it existed

richard-w commented on src/main.rs:
```diff
@@ -125,6 +140,49 @@ fn main() {
+            // Safe to unwrap, there will never be an empty string passed as an arg
+            let outer = files.next().unwrap();
+
+            let inner = match files.next() {
```
If you use `PathBuf` this block becomes unnecessary

richard-w commented on src/main.rs:
```diff
@@ -125,6 +140,49 @@ fn main() {
+                    outer.rsplit('/').next().unwrap()
+                }
+            };
+            let mut inner = inner.split('/');
```
This is not a safe way to split paths. Rather use `Path::components()`.

Something like this:
```rust
let mut inner_dir_components = inner.components().take(inner.components().count() - 1);                                                                                                                
match inner_dir_components.next() {                                                                                                                                                                    
    Some(std::path::Component::RootDir) => {},                                                                                                                                                         
    _ => panic!("Invalid --add-file argument: Inner path is not absolute"),                                                                                                                            
}                                                                                                                                                                                                      
let mut current_fs_dir = fs.root_dir();                                                                                                                                                                
for inner_dir_component in inner_dir_components {                                                                                                                                                      
    match inner_dir_component {                                                                                                                                                                        
        std::path::Component::Normal(dir_osstr) => {                                                                                                                                                   
            let dir_str = dir_osstr.to_str().expect("Invalid --add-file argument: Invalid inner path encoding");                                                                                       
            current_fs_dir = current_fs_dir.create_dir(dir_str).expect("Failed to create directory on EFI image");                                                                                     
        }                                                                                                                                                                                              
        _ => panic!("Invalid --add-file argument: Invalid inner path"),                                                                                                                                
    }                                                                                                                                                                                                  
} 
```

richard-w commented on src/main.rs:
```diff
@@ -125,6 +140,49 @@ fn main() {
+
+        // Create user provided additional files
+        for files in additional_files {
+            let mut user_file = fs.root_dir();
```
I would call this `current_fs_dir` and only define `user_file` when it's actually pointing to a file and not a directory.

richard-w commented on src/main.rs:
```diff
@@ -125,6 +140,49 @@ fn main() {
+            }
+            let mut user_file = user_file.create_file(inner_file).unwrap();
+
+            let data = match std::fs::read(outer) {
+                Ok(data) => data,
+                Err(e) => {
+                    eprint!("Error reading file: {}\n{}\n", outer, e);
+                    std::process::exit(1);
+                }
+            };
```
```suggestion
            let data = std::fs::read(outer).expect("Invalid --add-file argument: Failed to read outer file");
```

richard-w reviewed (changes requested):
Sorry for taking so long to review this. I usually try to be more responsive but I wasn't able to make time any sooner.

I think this PR is getting on track. Thank you for your work. I have added some suggestions to improve the code.

Also please squash your old commits so the history does not contain the binary file any more.

DjLogozzo commented:
Hey, I've implemented the changes you've requested, and squashed this Pull Request into one commit, Sorry for the massive gap between the changes requested, and the changes implemented, life got in the way and I haven't had the time or energy to contribute to coding.

If there are any further tweaks or changes you'd like, don't hesitate to ask

Richard-W commented:
Hi, Sorry for the long delay. I am currently pretty busy and do not have a lot of free time on my hands. I will try to review this soon, but it may take a while.

Something I already noticed: Can you edit your commit message to actually describe your feature? Something along the lines of the title of this PR? I losely follow GNOME conventions [1] usually.

[1]: https://wiki.gnome.org/Git/CommitMessages

DjLogozzo commented:
I've updated the most recent commit message, as well as the PR description (keeping the old one for historic purposes, if you'd like me to remove it completely I can).

Richard-W commented:
The format check fails. Please run `cargo fmt` on your changes.

richard-w reviewed (changes requested):
`cargo fmt` should be run.

DjLogozzo committed:
Formatted Code

Ran `cargo fmt` to format the code

Richard-W commented:
bors r+

Richard-W commented:
Thank you for your contribution.
